File: src/scrapers/inmuebles24.js
/**
 * Inmuebles24.com Scraper
 *
 * Scrapes real estate developer / agency listings from Mexico's largest
 * real estate portal. Extracts agency name, agent name, phone, email,
 * and location for the Riviera Maya corridor.
 *
 * URL pattern: https://www.inmuebles24.com/inmuebles-en-venta-en-{location}.html
 */

import { PlaywrightCrawler, RequestList } from 'crawlee';

const BASE_URL = 'https://www.inmuebles24.com';

const LOCATION_SLUGS = {
    'Playa del Carmen': 'playa-del-carmen',
    'Tulum':            'tulum',
    'Cancun':           'cancun',
    'Bacalar':          'bacalar',
    'Puerto Morelos':   'puerto-morelos',
    'Akumal':           'akumal',
    'Holbox':           'holbox',
    'Cozumel':          'cozumel',
    'Mahahual':         'mahahual',
};

/**
 * @param {{ locations: string[], maxLeadsPerSource: number }} opts
 * @returns {Promise<import('../utils/leads.js').Lead[]>}
 */
export async function scrapeInmuebles24({ locations, maxLeadsPerSource }) {
    const leads = [];

    const startUrls = locations
        .filter(loc => LOCATION_SLUGS[loc])
        .map(loc => ({
            url: `${BASE_URL}/inmuebles-en-venta-en-${LOCATION_SLUGS[loc]}.html`,
            userData: { location: loc, page: 1 },
        }));

    const crawler = new PlaywrightCrawler({
        maxRequestsPerCrawl: startUrls.length * 3,
        requestHandlerTimeoutSecs: 30,

        async requestHandler({ page, request, enqueueLinks }) {
            const { location } = request.userData;
            await page.waitForSelector('[data-qa="posting PROPERTY"]', { timeout: 15000 }).catch(() => {});

            const listings = await page.$$eval('[data-qa="posting PROPERTY"]', (cards) =>
                cards.map(card => {
                    const agency = card.querySelector('[data-qa="posting-card-publisher-name"]')?.textContent?.trim() ?? '';
                    const agent  = card.querySelector('[data-qa="posting-card-publisher-agent"]')?.textContent?.trim() ?? '';
                    const phone  = card.querySelector('[data-qa="posting-card-phone"]')?.textContent?.trim() ?? '';
                    const title  = card.querySelector('[data-qa="posting-card-title"]')?.textContent?.trim() ?? '';
                    const price  = card.querySelector('[data-qa="posting-card-price"]')?.textContent?.trim() ?? '';
                    const address = card.querySelector('[data-qa="posting-card-location"]')?.textContent?.trim() ?? '';
                    const link   = card.querySelector('a')?.href ?? '';
                    return { agency, agent, phone, title, price, address, link };
                })
            );

            for (const listing of listings) {
                if (!listing.agency && !listing.agent) continue;
                leads.push({
                    source: 'Inmuebles24',
                    businessName: listing.agency,
                    contactName: listing.agent,
                    email: '',
                    phone: listing.phone,
                    website: '',
                    address: listing.address,
                    city: location,
                    state: 'Quintana Roo',
                    country: 'Mexico',
                    category: 'Real Estate Agency',
                    rating: '',
                    reviewCount: 0,
                    googleMapsUrl: '',
                    description: `${listing.title} — ${listing.price}`,
                    listingUrl: listing.link,
                    tags: 'real-estate, inmuebles24',
                });
            }

            if (leads.length < maxLeadsPerSource) {
                const nextButton = await page.$('[data-qa="PAGING_NEXT"]');
                if (nextButton) {
                    await enqueueLinks({ selector: '[data-qa="PAGING_NEXT"]' });
                }
            }
        },

        failedRequestHandler({ request, error }) {
            console.warn(`  [Inmuebles24] Failed: ${request.url} — ${error.message}`);
        },
    });

    const requestList = await RequestList.open(null, startUrls);
    await crawler.run(startUrls.map(r => r.url));

    return dedup(leads).slice(0, maxLeadsPerSource);
}

function dedup(leads) {
    const seen = new Set();
    return leads.filter(l => {
        const key = `${l.businessName}|${l.contactName}|${l.city}`.toLowerCase();
        if (seen.has(key)) return false;
        seen.add(key);
        return true;
    });
}

File: recrea_scrapy/recrea_scrapy/spiders/inmuebles24.py
"""
Inmuebles24.com Spider — Real estate agencies & agents in Riviera Maya
Uses scrapy-playwright to bypass Cloudflare/403 blocks.
"""
import scrapy
from scrapy_playwright.page import PageMethod
from recrea_scrapy.items import LeadItem

LOCATION_SLUGS = {
    "Playa del Carmen": "playa-del-carmen",
    "Tulum":            "tulum",
    "Cancun":           "cancun",
    "Bacalar":          "bacalar",
    "Puerto Morelos":   "puerto-morelos",
    "Akumal":           "akumal",
    "Holbox":           "holbox",
    "Cozumel":          "cozumel",
    "Mahahual":         "mahahual",
}

class Inmuebles24Spider(scrapy.Spider):
    name = "inmuebles24"
    base_url = "https://www.inmuebles24.com"

    def start_requests(self):
        for city, slug in LOCATION_SLUGS.items():
            url = f"{self.base_url}/inmuebles-en-venta-en-{slug}.html"
            yield scrapy.Request(
                url,
                callback=self.parse,
                meta={
                    "playwright": True,
                    "playwright_context": "default",
                    "playwright_page_methods": [
                        PageMethod("wait_for_load_state", "networkidle"),
                        PageMethod(
                            "wait_for_selector",
                            '[data-qa="posting PROPERTY"], .posting-card, article',
                            timeout=20000,
                        ),
                    ],
                    "city": city,
                },
                errback=self.errback,
            )

    def parse(self, response):
        city = response.meta["city"]

        for card in response.css('[data-qa="posting PROPERTY"], .posting-card, article.avisoNormal'):
            agency  = card.css('[data-qa="posting-card-publisher-name"], .publisher-name').css('::text').get('').strip()
            agent   = card.css('[data-qa="posting-card-publisher-agent"], .agent-name').css('::text').get('').strip()
            phone   = card.css('[data-qa="posting-card-phone"], [href^="tel:"]').css('::text').get('').strip()
            title   = card.css('[data-qa="posting-card-title"], h2, h3').css('::text').get('').strip()
            price   = card.css('[data-qa="posting-card-price"], .price').css('::text').get('').strip()
            address = card.css('[data-qa="posting-card-location"], .address').css('::text').get('').strip()
            link    = card.css('a::attr(href)').get('')

            if not agency and not agent:
                continue

            item = LeadItem()
            item['source']       = 'Inmuebles24'
            item['businessName'] = agency
            item['contactName']  = agent
            item['email']        = ''
            item['phone']        = phone
            item['website']      = ''
            item['address']      = address
            item['city']         = city
            item['state']        = 'Quintana Roo'
            item['country']      = 'Mexico'
            item['category']     = 'Real Estate Agency'
            item['rating']       = ''
            item['description']  = f"{title} — {price}".strip(' — ')
            item['listingUrl']   = link if link.startswith('http') else f"{self.base_url}{link}"
            item['googleMapsUrl'] = ''
            item['tags']         = 'real-estate, inmuebles24'
            yield item

        # Pagination
        next_page = response.css('[data-qa="PAGING_NEXT"] a::attr(href), a[rel="next"]::attr(href)').get()
        if next_page:
            yield response.follow(
                next_page,
                self.parse,
                meta={
                    "playwright": True,
                    "playwright_context": "default",
                    "playwright_page_methods": [
                        PageMethod("wait_for_load_state", "networkidle"),
                    ],
                    "city": city,
                },
                errback=self.errback,
            )

    def errback(self, failure):
        self.logger.error(f"[Inmuebles24] Request failed: {failure.request.url} — {failure.value}")
